Zoooooooom

BlueBazze opened the pull request:
## Issue
Generally missing a zoom functionality.
Mentioned in:
https://github.com/apivideo/api.video-reactnative-live-stream/issues/9

## Description of the Change
Added property "zoomEnabled" that defines if the pinch gesture handler should be active. The actual zoom function is called "Zoom" which uses the property currentZoomFactor to keep a persistant zoom factor across multiple gesture frames.

At the moment the zoom functionality will not work until a stream has been started. - This will have to be fixed before being merged. But need to know how this should be handled as to not break any other libraries depending on this one.

## Alternate Designs
It would be possible to have a gesture handler directly from React Native or Flutter which just passes down a zoom factor which will be passed into `rtmpStream.setZoomFactor`. But in my experience is not ideal. You have two different systems that output the pinch gesture data slightly different. One method would give you two different outcomes.

## Possible Drawbacks
Some people might get confused. With React Native, if you draw a View with Position: "Absolute" over the preview. That View would capture all of the click events. So you will need to add `pointerEvents="none"` to the view stretching over the preview to make touches go through that view.

## Verification Process
Tested in development mode with both the native example given by this library. And also tested with example given with the api.video-reactnative-live-stream library.

Asserted that the zoom did not function until the stream were started.

BlueBazze commented:
The `zoomEnabled = true/false` lines will be removed, they are for testing purposes.

As mentioned.
Gotta know if any of this would break anything in the Flutter package? And if you have any changes to this.
Havent implemented this with Android yet. But some work has already been done - https://github.com/ThibaultBee/StreamPack/commit/4b2d37e7b92e59b4aefc6e0c0e586e13140310a0

Do you have any changes?

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -17,6 +17,11 @@ public class ApiVideoLiveStream {
     public var onDisconnect: (() -> Void)?
 
+    //Gotta be initialized before init, otherwise .self wont exist in init - Aka voodoo
+    private var zoomGesture: UIPinchGestureRecognizer = UIPinchGestureRecognizer()
```
use `lazy var` instead of workaround

SyncronetDev replied:
Done

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -17,6 +17,11 @@ public class ApiVideoLiveStream {
+    //Gotta be initialized before init, otherwise .self wont exist in init - Aka voodoo
+    private var zoomGesture: UIPinchGestureRecognizer = UIPinchGestureRecognizer()
+    private var currentZoomFactor: CGFloat = 1.0
+    private var lastScale: CGFloat = 0.0
```
unused -> remove

SyncronetDev replied:
Done

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -64,6 +69,19 @@ public class ApiVideoLiveStream {
     }
 
+    // TODO: Fix zoom flow so it can be a React prop instead of method
+    public var zoomEnabled: Bool = false {
```
maybe rename to `pinchedZoomEnabled` instead.
We might want to directly have a `setZoomFactor` API. `zoomEnabled` won't have sense in this case.

SyncronetDev replied:
Done

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -124,6 +142,9 @@ public class ApiVideoLiveStream {
                 maxWidth, maxHeight, width, height, centerX, centerY,
             ])
+
+            // TODO: Fix zoom flow so it can be a React prop instead of method
```
Remove TODO

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -189,12 +210,18 @@ public class ApiVideoLiveStream {
         rtmpConnection.addEventListener(.ioError, selector: #selector(rtmpErrorHandler), observer: self)
 
+        // TODO: Fix zoom flow so it can be a React prop instead of method
+        zoomEnabled = true
```
To be remove (I know you know)

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -189,12 +210,18 @@ public class ApiVideoLiveStream {
     /// - Returns: Void
     public func stopStreaming() {
+        // TODO: Fix zoom flow so it can be a React prop instead of method
+        zoomEnabled = true
```
To be remove (I know you know)

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -250,6 +277,17 @@ public class ApiVideoLiveStream {
+    private func zoom(sender: UIPinchGestureRecognizer) {
+        if sender.state == .began || sender.state == .changed {
+            if(sender.scale < 2 && sender.scale > 0) {
+                currentZoomFactor = min(max(currentZoomFactor + (min(max(sender.scale, 0.5), 1.5) - 1) * 2.2, 0), 4)
```
Have you read this line:
https://github.com/shogo4405/HaishinKit.swift/blob/64571447d123be0a92ab58562a0c183a3799e0db/Platforms/iOS/AVVideoIOUnit%2BExtension.swift#L15

I guess you don't have to set limit to 0 and 4. HK already does this.

Anyway, I am not sure to understand the formula. Can you explain?

SyncronetDev replied:
The outer min-max is a boundary for the total zoom factor. Yes, it is to an extend set by HK. But here we are setting a local variable, meaning HK does not clamp our currentZoomFactor. So having our own clamp as a measure for the user that keeps zooming. Without this clamp you can zoom to a factor of 100 or -100, meaning you will have to zoom all the way towards 0 again before anything is affected again. 

currentZoomFactor in the calculation is the previous zoom factor.

The inner min-max is a boundary for how much you can affect the zoom in a single frame. (This is just to ignore outlier data - It is fairly hard to go past 0.9 or 1.1, but it can happen that the user somehow tricks the phone into thinking you moved a finger from one corner to the other in a single frame) - This is not a necessary, can be removed.

`sender.scale`: is the ratio change in the gesture compared to the last frame. We minus this value by 1 in order to magnify the scale, and it makes it so we can just add it to the previous zoom factor value (currentZoomFactor)
The `* 2.2` is simply to magnify the zoom. If this wasnt present the zoom would feel very slow.

Simplified: 
We take the pinch scale minus 1 (Above 0 == zoom in | Below 0 == zoom out), we multiply this value by a multiplier to magnify the effect of the pinch action. Then we take that value and add it to the previous zoom factor.
Now this sum value is the new zoom factor which we save in currentZoomFactor & apply this to HK zoom factor.
Afterwards we reset the `sender.scale` to 1 so it is ready for the next frame.

I did try to set sender.scale = 0, but that just stops the pinch event.

SyncronetDev replied:
The ramping & withRate. Did try them out, and the zooming felt slow. But could make a zoomConfig that includes the options ramp, withRate, pinchZoomMultiplier & maxZoom.

ThibaultBee replied:
The only thing that trigger me, is that if the device has more a maximum zoom > 4, the user won't be able to access his maximum zoom. Moreover, the minimum zoom is 1.0, so when the `currentZoomFactor` is going from 0 to 1, nothing happens for the user till the zoom factor is above 1.0.

A way to deal with this, could be to not clamp the `currentZoomFactor` and then use HK `zoomFactor` API (https://github.com/shogo4405/HaishinKit.swift/blob/64571447d123be0a92ab58562a0c183a3799e0db/Platforms/iOS/AVVideoIOUnit%2BExtension.swift#L7) like this:
```
let zoomFactor = currentZoomFactor + (sender.scale-1) * pinchZoomMultiplier
rtmpStream.setZoomFactor(currentZoomFactor)
currentZoomFactor = rtmpStream.zoomFactor
```

The other way is to clamp with the real value from `1.0` to `device.activeFormat.videoMaxZoomFactor`

ThibaultBee commented on Sources/ApiVideoLiveStream.swift:
```diff
@@ -64,6 +69,19 @@ public class ApiVideoLiveStream {
         }
     }
 
+    // TODO: Fix zoom flow so it can be a React prop instead of method
```
Comment to be removed

ThibaultBee commented:
Hi,

Thank for the PR :)

I am not great at RN but I was thinking on doing it the same way. 
Have in mind that someday we might add a `setZoomFactor` API (for zoom slider).

You shouldn't mind of the api.video Flutter package as it does not depends on this library anymore.

There is an easy zoom factor API in StreamPack but  it is only supported for Android >= 11. Before Android 11, zoom API is garbage. Is it ok with you?

BlueBazze commented:
Some zoom is better than no zoom.
Is there any deeper reason to why zoom wouldn't work on Android versions below android 11?

Can imagine that people using this kind of package would upgrade to newer phones when using it regularly.

Took a look at android, the implementation seems to be very similar to this pr. I'll finish it one of the following days along with a React Native pr to hook it all up.
Gonna include the setZoomFactor method you talked about.
Zoom enabled is gonna be renamed to NativeZoomEnabled. - People can disable that to use their own feature for zoom.
Hopefully should be all done and tested by Monday.

ThibaultBee commented:
> Some zoom is better than no zoom. Is there any deeper reason to why zoom wouldn't work on Android versions below android 11?

The answer is simple, it is because I didn't implement the zoom factor for version < 11.

BlueBazze commented:
Removed the zoom gesture, added it to the example instead. As to let users configure zoom gestures by themselves.
Discussed in https://github.com/apivideo/api.video-android-live-stream/pull/9

ThibaultBee commented:
Yes! Great job :)
Thank you! 👍
